### ISSUE TITLE
Using customize sampler in DDP TypeError: __init__() missing 2 required positional arguments

### ISSUE THREAD

**Issue by di0002ya** [Lightning-AI/lightning#12689]
## 🐛 Bug

I tried to use following customised sampler. Code worked well in single gpu. However, it returns type error when I train using 8 gpus with ddp. May I know how to solve it?  
```py

import itertools

import numpy as np
 
import tqdm

def safe_random_choice(input_data, size):
    replace = len(input_data) < size
    return np.random.choice(input_data, size=size, replace=replace)
 
class HierarchicalSampler:
    def __init__(
        self,
        dataset,
        batch_size,
        samples_per_class,
        batches_per_super_pair,
        nb_categories=2,
        drop_last = False
    ): 
        self.batch_size = int(batch_size)
        self.batches_per_super_pair = int(batches_per_super_pair)
        self.samples_per_class = int(samples_per_class)
        self.nb_categories = int(nb_categories)
        self.ds = dataset
        self.drop_last = drop_last
        # checks
        assert self.batch_size % self.nb_categories == 0, f"batch_size should be a multiple of {self.nb_categories}"
        self.sub_batch_len = self.batch_size // self.nb_categories
        if self.samples_per_class > 0:
            assert self.sub_batch_len % self.samples_per_class == 0, "batch_size not a multiple of samples_per_class"
        else:
            self.samples_per_class = None

        self.super_image_lists = dataset.super_dict.copy()
        self.super_pairs = list(itertools.combinations(set(dataset.super_labels), self.nb_categories))
        self.reshuffle()

    
    def __iter__(self,):
        self.reshuffle()
        for batch in self.batches:
            yield batch

    def __len__(self,):
        return len(self.batches)

    def __repr__(self,):
        return (
            f"{self.__class__.__name__}(\n"
            f"    batch_size={self.batch_size},\n"
            f"    samples_per_class={self.samples_per_class},\n"
            f"    batches_per_super_pair={self.batches_per_super_pair},\n"
            f"    nb_categories={self.nb_categories}\n)"
        )

    def reshuffle(self):
        batches = []
        for combinations in tqdm.tqdm(self.super_pairs):

            for b in range(self.batches_per_super_pair):

                batch = []
                for slb in combinations:

                    sub_batch = []
                    all_classes = list(self.super_image_lists[slb].keys())
                    np.random.shuffle(all_classes)
                    for cl in all_classes:
                        instances = self.super_image_lists[slb][cl]
                        samples_per_class = self.samples_per_class if self.samples_per_class else len(instances)
                        if len(sub_batch) + samples_per_class > self.sub_batch_len:
                            continue
                        sub_batch.extend(safe_random_choice(instances, size=samples_per_class))

                    batch.extend(sub_batch)

                np.random.shuffle(batch)
                batches.append(batch)

        np.random.shuffle(batches)
        self.batches = batches


```

Error output is 
Error executing job with overrides: ['trainer.gpus=8', '+trainer.strategy=ddp']

```
Traceback (most recent call last):
  File "/workspace/jupyter_notebooks/mavis_personal_repo/instance-img-retrieval/Dev/train.py", line 22, in main
    return train(config)
  File "/workspace/jupyter_notebooks/mavis_personal_repo/instance-img-retrieval/Dev/src/training_pipeline.py", line 127, in train
    trainer.fit(model=model, datamodule=datamodule)
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 771, in fit
    self._call_and_handle_interrupt(
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 724, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 812, in _fit_impl
    results = self._run(model, ckpt_path=self.ckpt_path)
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1237, in _run
    results = self._run_stage()
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1324, in _run_stage
    return self._run_train()
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1354, in _run_train
    self.fit_loop.run()
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/loops/base.py", line 199, in run
    self.on_run_start(*args, **kwargs)
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/loops/fit_loop.py", line 209, in on_run_start
    self.trainer.reset_train_val_dataloaders(self.trainer.lightning_module)
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 2004, in reset_train_val_dataloaders
    self.reset_train_dataloader(model=model)
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 1871, in reset_train_dataloader
    self.train_dataloader = apply_to_collection(
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/utilities/apply_func.py", line 99, in apply_to_collection
    return function(data, *args, **kwargs)
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/trainer/connectors/data_connector.py", line 300, in _prepare_dataloader
    dataloader = _update_dataloader(dataloader, sampler, mode=mode)
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/utilities/data.py", line 176, in _update_dataloader
    dl_kwargs = _get_dataloader_init_kwargs(dataloader, sampler, mode=mode)
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/utilities/data.py", line 230, in _get_dataloader_init_kwargs
    dl_kwargs.update(_dataloader_init_kwargs_resolve_sampler(dataloader, sampler, mode=mode))
  File "/workspace/miniconda3/envs/firstEnv/lib/python3.8/site-packages/pytorch_lightning/utilities/data.py", line 286, in _dataloader_init_kwargs_resolve_sampler
    batch_sampler = type(batch_sampler)(
TypeError: __init__() missing 2 required positional arguments: 'samples_per_class' and 'batches_per_super_pair'


```

cc @justusschock @awaelchli @ninginthecloud @rohitgr7

**Comment by di0002ya** [comment]
Hi team, after reading documents, I set up `trainer.replace_sampler_ddp = False` and process worked. But in other issues, I see supporter reminds the user to use distributedSampler. Do I need to use it as well?

**Comment by awaelchli** [comment]
Hey @di0002ya 
We should provide a better error message there that hints at the Trainer argument. 
Note, if you don't use the default distributed sampler from pytorch, you need to add the relevant logic directly into your custom sampler. Otherwise, your data will be duplicated across your devices instead of sharded. You can look at the implementation of the torch [DistributedSampler](https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler).

**Comment by di0002ya** [comment]
> 

Hey @awaelchli, thanks for your guidance. I will try to do it.

**Comment by awaelchli** [comment]
You are welcome. I will keep the issue open because I think you brought awareness to the issue of our error messaging, something we can improve on here.

**Comment by carmocca** [comment]
I believe this has been fixed by @otaj in the 1.7 release

**Comment by otaj** [comment]
Hi @carmocca, @di0002ya. This was _almost_ solved in the 1.7 release. The only thing that's missing is for user code to subclass `BatchSampler` from regular PyTorch, otherwise our magic won't work :zap: And, since it's a `BatchSampler`, we still require you to use underlying `Sampler`. But yes, in general, it has been fixed :muscle:
